### ISSUE TITLE
libzfs_core.lzc_get_props() returns wrong value for parent dataset if child datasets are present

### ISSUE THREAD

**Issue by spamferkel** [ClusterHQ/pyzfs#27]
If `libzfs_core.lzc_get_props(b'parent-dataset')` is executed, it returns properties for the first child dataset instead of the parents own properties. If more than one child is present it returns the properties of the first child listed by libzfs_core.lzc_list_children(b'parent-dataset').

My environment is running python2.7 with pyzfs-0.2.2 and zfs-0.6.5.4-r2 on Gentoo.

Best regards

**Comment by avg-I** [comment]
There could have been a change in `lzc_list()` behavior that pyzfs hasn't accounted for yet.
@ryao could you please take a look if possible?
The uncommitted part of the API (lzc_ calls that are not in illumos yet) could still be unstable.

BTW, you could run `python -m unittest -v libzfs_core.test.test_libzfs_core` to check for problems.

**Comment by spamferkel** [comment]
the unittest results in a segfault:

```
$> sudo python2 -m unittest -v libzfs_core.test.test_libzfs_core > unittest_results
test_bookmarks (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_cross_pool (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_duplicate_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_empty (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_for_the_same_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_for_the_same_snap_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_invalid_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_invalid_name_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_mismatching_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_mismatching_names (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_missing_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_missing_snaps (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_partially_mismatching_names (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_too_long_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_bookmarks_too_long_name_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_clone (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_clone_across_pools (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_clone_across_pools_to_ro_pool (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_clone_invalid_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_clone_invalid_pool_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_clone_invalid_snap_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_clone_invalid_snap_name_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_clone_nonexistent_parent_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_clone_nonexistent_snapshot (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_clone_to_nonexistent_pool (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_below_zvol (libzfs_core.test.test_libzfs_core.ZFSTest) ... skipped 'https://www.illumos.org/issues/6101'
test_create_fs_duplicate (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_in_nonexistent_pool (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_in_ro_pool (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_with_invalid_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_with_invalid_pool_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_with_invalid_prop (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_with_invalid_prop_type (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_with_invalid_prop_val (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_with_prop (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_without_parent (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_fs_wrong_ds_type (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_create_zvol (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_deferred_destroy_cloned_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy_bookmark_nonexistent_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy_bookmarks (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy_bookmarks_empty (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy_bookmarks_invalid_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy_cloned_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy_invalid_snap_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... expected failure
test_destroy_mounted_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... skipped 'snapshot is not auto-unmounted'
test_destroy_nonexistent (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy_nonexistent_snapshot (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy_snapshot_of_nonexistent_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy_snapshot_of_nonexistent_pool (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_destroy_too_long_full_snap_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... expected failure
test_destroy_too_long_short_snap_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_exists (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_exists_failure (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_exists_in_ro_pool (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_full_already_existing_with_same_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_full_already_existing_with_snapshots (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_full_existing_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_full_existing_modified_mounted_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_full_missing_parent_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_incremental_modified_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_incremental_modified_fs_plus_cloned_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_incremental_modified_fs_plus_held_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_incremental_modified_fs_plus_later_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_incremental_modified_fs_plus_same_name_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_force_recv_incremental_modified_mounted_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_bookmarks (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_bookmarks_invalid_property (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_bookmarks_nonexistent_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_fs_props (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_holds (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_holds_after_auto_cleanup (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_holds_invalid_snap_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_holds_invalid_snap_name_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... expected failure
test_get_holds_nonexistent_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_holds_too_long_snap_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_holds_too_long_snap_name_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_mountpoint_legacy (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_mountpoint_none (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_mountpoint_path (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_props_nonexistent (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_snap_clones (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_get_snap_props (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_across_pools (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_bad_fd (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_bad_fd_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_bad_fd_3 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_duplicate (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_empty (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_empty_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_fd (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_invalid_snap_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_invalid_snap_name_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_many_snaps (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_many_tags (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_many_with_all_missing (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_many_with_one_missing (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_missing_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_missing_fs_auto_cleanup (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_too_long_snap_name (libzfs_core.test.test_libzfs_core.ZFSTest) ... expected failure
test_hold_too_long_snap_name_2 (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_too_long_tag (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_vs_snap_destroy (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_hold_wrong_fd (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_inherit_missing_prop (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_inherit_prop (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_inherit_prop_on_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_inherit_readonly_prop (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_inherit_unknown_prop (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_list_children (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_list_children_nonexistent (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_list_children_of_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_list_snaps (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_list_snaps_nonexistent (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_list_snaps_of_snap (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_multiple_existing_snapshots (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_multiple_new_and_existing_snapshots (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_multiple_snapshots (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_multiple_snapshots_for_same_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_multiple_snapshots_multiple_nonexistent_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_multiple_snapshots_nonexistent_fs (libzfs_core.test.test_libzfs_core.ZFSTest) ... ok
test_promote (libzfs_core.test.test_libzfs_core.ZFSTest) ... Speicherzugriffsfehler
```

**Comment by avg-I** [comment]
Thank you for testing. Sorry for the pause, I am going to look into this tomorrow.

**Comment by avg-I** [comment]
I've been able to reproduce the issue, working on a fix.

**Comment by avg-I** [comment]
Should be fixed by PR #30
